transport/httptransport/libp2p_server.go

```diff
@@ -0,0 +1,620 @@
+	content     *car.CarReaderSeeker
+	// indicates whether this transfer replaces a previous transfer with the
+	// same id
+	isRestart bool
+
+	lk     sync.RWMutex
+	err    error
+	sent   uint64
+	status types.TransferStatus
+
+	// Set when the transfer is replaced with a new transfer with the same id.
+	// This prevents the replaced transfer from firing any more events.
+	closed bool
+
+	// When the error or complete event has been fired
+	eventsDrained chan struct{}
+}
+
+func newLibp2pTransfer(val *AuthValue, authToken string, localAddr string, remoteAddr string, content *car.CarReaderSeeker) *Libp2pTransfer {
+	return &Libp2pTransfer{
+		ID:            val.ID,
+		PayloadCid:    val.PayloadCid,
+		ProposalCid:   val.ProposalCid,
+		CreatedAt:     time.Now(),
+		AuthToken:     authToken,
+		LocalAddr:     localAddr,
+		RemoteAddr:    remoteAddr,
+		content:       content,
+		status:        types.TransferStatusStarted,
+		eventsDrained: make(chan struct{}),
+	}
+}
+
+// setSent is called when some bytes are sent.
+// Returns the state of the transfer after the sent amount is updated.
+func (t *Libp2pTransfer) setSent(sent uint64) types.TransferState {
+	t.lk.Lock()
+	defer t.lk.Unlock()
+
+	t.sent = sent
+	t.status = types.TransferStatusOngoing
+
+	return t.state()
+}
+
+// setComplete is called when the transfer completes or there is an error
+// Returns the state of the transfer after the complete status is applied.
+func (t *Libp2pTransfer) setComplete(err error) types.TransferState {
+	t.lk.Lock()
+	defer t.lk.Unlock()
+
+	t.err = err
+	if t.err == nil {
+		t.status = types.TransferStatusCompleted
+	} else {
+		t.status = types.TransferStatusFailed
+	}
+
+	return t.state()
+}
+
+func (t *Libp2pTransfer) State() types.TransferState {
+	t.lk.RLock()
+	defer t.lk.RUnlock()
+
+	return t.state()
+}
+
+func (t *Libp2pTransfer) state() types.TransferState {
+	msg := ""
+	if t.err != nil {
+		msg = t.err.Error()
+	} else {
+		msg = fmt.Sprintf("transferred %d bytes", t.sent)
+		if t.sent == 0 {
+			msg = "pull data transfer queued"
+		}
+	}
+	return types.TransferState{
+		LocalAddr:  t.LocalAddr,
+		RemoteAddr: t.RemoteAddr,
+		Status:     t.status,
+		Sent:       t.sent,
+		Message:    msg,
+		PayloadCid: t.PayloadCid,
+	}
+}
+
+// cancel permanently fails the transfer, and waits for the error or completed
+// event to be fired before returning.
+// It returns the state that the transfer is in after being canceled (either
+// completed or errored out).
+func (t *Libp2pTransfer) cancel(ctx context.Context) (*types.TransferState, error) {
+	// Cancel the read / write stream
+	err := t.content.Cancel(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Wait for the error or completed event to be fired
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-t.eventsDrained:
+		t.lk.Lock()
+		defer t.lk.Unlock()
+		st := t.state()
+		return &st, nil
+	}
+}
+
+func (t *Libp2pTransfer) close(ctx context.Context) {
+	t.closed = true
```

Comment:
We should also take the `Libp2pTransfer ` lock here before mutating this field.

Reply:
t.closed is only modified and accessed from the transfersMgr event loop, so it doesn't need to be under a lock